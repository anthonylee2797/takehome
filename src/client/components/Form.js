import React, { useState } from 'react'
import PropTypes from 'prop-types'
import Api from '../api/api'

const Form = ({comments, setComments}) => {
  const [input, setInput] = useState({name: "", comment: ""})

  async function addComment (e) {
    e.preventDefault()

    const { name , comment } = {...input}

    if (name.length === 0 || comment.length === 0) {
      alert('name or message not inputted correctly')
      return
    }

    // clearing input
    setInput({name: "", comment: ""})

    // posting to database and returning comment to client
    try {
      const commentID = await Api.post('/createComment', { name, message: comment })
      const addedComment = await Api.get(`/getComment/?id=${commentID.id}`)

      // setting state of comments
      const newComments = [addedComment, ...comments]
      setComments(newComments)
    } catch (error) {
      console.log(error)
    }
  }

  function handleNameInput (e) {
    const name = e.target.value
    setInput({...input, name})
  }

  function handleCommentInput (e) {
    const comment = e.target.value
    setInput({...input, comment})
  }

  return (
    <div>
      <form className="form" onSubmit={addComment}>
        <div  className='form-name'>
          <label htmlFor='name'>Name</label>
          <input name="name" onChange={handleNameInput} value={input.name} placeholder="Insert Name" />
        </div>

        <div className='form-comment'>
          <label htmlFor='comment'>Comment</label>
          <textarea name="comment" onChange={handleCommentInput} value={input.comment}  placeholder="Insert Comment" />
        </div>
        <button data-testid="submit-btn" className='form-button' type="submit">Add Comment</button>
      </form>
    </div>
  )
}

Form.propTypes = {
  comments: PropTypes.array,
  setComments: PropTypes.func
}

export default Form
