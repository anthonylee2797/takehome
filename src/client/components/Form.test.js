import React from 'react'
import Form from './Form'
import { render, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom'

const setup = () => {
  const utils = render(<Form />)
  const nameInput = utils.getByPlaceholderText('Insert Name')
  const commentInput = utils.getByPlaceholderText('Insert Comment')
  const submit = utils.getByTestId('submit-btn')

  return {
    submit,
    nameInput,
    commentInput,
    ...utils
  }
}

test('should render form correctly', () => {
  const form = render(<Form />)
  expect(form.container).toMatchSnapshot()
})

test('should not submit form when no values in name input or message text area', () => {
  const { nameInput, commentInput, submit } = setup()
  global.alert = jest.fn()

  fireEvent.change(nameInput, { target: { value: '' } })
  fireEvent.change(commentInput, { target: { value: '' } })
  fireEvent.click(submit)
  expect(global.alert).toHaveBeenCalledTimes(1)
})

test('should submit comment and reset input values upon submit', () => {
  const { nameInput, commentInput, submit } = setup()
  fireEvent.change(nameInput, { target: { value: 'name' } })
  fireEvent.change(commentInput, { target: { value: 'comment' } })
  fireEvent.click(submit)

  expect(nameInput.value).toBe('')
  expect(commentInput.value).toBe('')
})
