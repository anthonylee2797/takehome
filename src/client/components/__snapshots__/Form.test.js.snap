// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render form correctly 1`] = `
<div>
  <div>
    <form
      class="form"
    >
      <div
        class="form-name"
      >
        <label
          for="name"
        >
          Name
        </label>
        <input
          name="name"
          placeholder="Insert Name"
          value=""
        />
      </div>
      <div
        class="form-comment"
      >
        <label
          for="comment"
        >
          Comment
        </label>
        <textarea
          name="comment"
          placeholder="Insert Comment"
        />
      </div>
      <div
        class="form-buttons"
      >
        <button
          class="form-button"
          data-testid="submit-btn"
          type="submit"
        >
          Add Comment
        </button>
        <button
          class="form-button"
        >
          Delete Comments
        </button>
      </div>
    </form>
  </div>
</div>
`;
